torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1362,6 +1362,160 @@ def rbinary_magic_impl(self, other):
+                "<=": operator.le,
+                "<": operator.lt,
+            }
+            assert isinstance(expr, (sympy.Eq, sympy.Ne, sympy.Ge, sympy.Gt, sympy.Le, sympy.Lt))
+            assert expr.rel_op in opmap, (
+                f"sympy relop not supported: {expr.rel_op}"
+            )
+            return opmap[expr.rel_op](self(expr.lhs), self(expr.rhs))
+
+        def _Add(self, expr: sympy.Add):
+            return sum(self(expr.args))
+
+        def _Mod(self, expr: sympy.Mod):
+            p, q = [self._to_int(a) for a in self(expr.args)]
+            return p % q
+
+        def _Mul(self, expr: sympy.Mul):
+            numerator_list = []
+            denominator_list = []
+
+            for item in expr.args:
+                if item.is_commutative and isinstance(item, sympy.Pow) and item.exp.is_Rational and item.exp.is_negative:
+                    denominator_list.append(sympy.Pow(item.base, -item.exp))
+                else:
+                    numerator_list.append(item)
+
+            numerator = self._int(1) if len(numerator_list) == 0 else math.prod(self(numerator_list))
+
+            if len(denominator_list) > 0:
+                denominator = math.prod(self(denominator_list))
+                return self._div(numerator, denominator)
+
+            return numerator
+
+        def _Pow(self, expr: sympy.Pow):
+            base, exp = self(expr.args)
+            return base ** exp
+
+        def _FloorDiv(self, expr: FloorDiv):
+            base, divisor = self(expr.args)
+            return base / divisor
+
+        def __call__(self, expr):
+            if isinstance(expr, (tuple, list)):
+                return type(expr)(self(e) for e in expr)
+            if isinstance(expr, sympy.Rel):
+                typename = "Rel"
+            elif isinstance(expr, sympy.Integer):
+                typename = "int"
+                expr = int(expr)
+            elif isinstance(expr, sympy.core.numbers.IntegerConstant):
+                constantmap = {
+                    sympy.S.Zero: 0,
+                    sympy.S.One: 1,
+                    sympy.S.NegativeOne: -1,
+                }
+                typename = "int"
+                expr = constantmap[expr]
+            else:
+                typename = type(expr).__name__
+
+            method = f"_{typename}"
+            assert hasattr(self, method), (
+                f"unsupported sympy operation: {expr} ({method})"
+            )
+            return getattr(self, method)(expr)
+
+    class TranslationValidator:
+        def __init__(self) -> None:
+            super().__init__()
+            self._symbols = {}
+            self._inputs = set()
+            self._outputs = set()
+            self._z3 = SympyToZ3(self._symbols)
+
+        def _add_freesymbols(self, e: sympy.Expr) -> None:
+            for s in e.free_symbols:
+                if s not in self._symbols:
+                    assert isinstance(s, sympy.Symbol) and s.is_integer and s.is_positive
```

Comment:
This assert won't always be true btw ;)

Reply:
Hmm. I left them there because I didn't see any failures. What part won't always be true? Everything?